src/op.rs

```diff
@@ -0,0 +1,2691 @@
+                                format);
+        }
+    }
+}
+
+#[test]
+fn test_decode_implicit_value() {
+    // Doesn't matter for this test.
+    let address_size = 4;
+    let format = Format::Dwarf32;
+
+    let data = b"hello";
+    let mut buffer = vec![constants::DW_OP_implicit_value.0];
+    leb128::write::unsigned(&mut buffer, data.len() as u64).unwrap();
+    buffer.write_all(data).unwrap();
+
+    // Too short.
+    for i in 1..buffer.len() - 1 {
+        check_decode_failure(&buffer[..i], Error::UnexpectedEof, address_size, format)
+    }
+
+    check_simple_decode(&buffer[..],
+                        &Operation::ImplicitValue { data: &data[..] },
+                        address_size,
+                        format);
+}
+
+/// A DWARF expression evaluation.
+pub struct Evaluation<'context, 'input, Endian>
+    where Endian: 'context + Endianity,
+          'input: 'context
+{
+    bytecode: &'input [u8],
+    address_size: u8,
+    format: Format,
+    initial_value: Option<u64>,
+    object_address: Option<u64>,
+    callbacks: &'context mut EvaluationContext<'input>,
+    max_iterations: Option<u32>,
+    phantom: PhantomData<Endian>,
+}
+
+impl<'context, 'input, Endian> Evaluation<'context, 'input, Endian>
+    where Endian: 'context + Endianity,
+          'input: 'context
+{
+    /// Create a new DWARF expression evaluator.
+    pub fn new(bytecode: &'input [u8],
+               address_size: u8,
+               format: Format,
+               callbacks: &'context mut EvaluationContext<'input>)
+               -> Evaluation<'context, 'input, Endian> {
+        Evaluation::<'context, 'input, Endian> {
+            bytecode: bytecode,
+            address_size: address_size,
+            format: format,
+            initial_value: None,
+            object_address: None,
+            callbacks: callbacks,
+            max_iterations: None,
+            phantom: PhantomData,
+        }
+    }
+
+    /// Set an initial value to be pushed on the DWARF expression
+    /// evaluator's stack.  This can be used in cases like
+    /// `DW_AT_vtable_elem_location`, which require a value on the
+    /// stack before evaluation commences.
+    pub fn set_initial_value(&mut self, value: u64) {
+        self.initial_value = Some(value);
+    }
+
+    /// Set the enclosing object's address, as used by
+    /// `DW_OP_push_object_address`
+    pub fn set_object_address(&mut self, value: u64) {
+        self.object_address = Some(value);
+    }
+
+    /// Set the maximum number of iterations to be allowed by the
+    /// expression evaluator.  The default is `None`.  This value can
+    /// be set to avoid denial of service attacks by bad DWARF
+    /// bytecode.
+    pub fn set_max_iterations(&mut self, value: u32) {
+        self.max_iterations = Some(value);
+    }
+
+    /// Evaluate a DWARF expression.
+    pub fn evaluate(&mut self) -> ParseResult<Vec<Piece<'input>>>
+        where Endian: Endianity
```

Comment:
This function has complicated control flow. Generally that means it's better to split it up into a few functions. In this case though, that would be difficult due to the amount of state needed. Perhaps the state could be moved into a helper struct? Just something to consider, feel free to ignore/defer.